doc/rst/tools/chplcheck/chplcheck.rst

```diff
@@ -0,0 +1,288 @@
+.. _readme-chplcheck:
+
+chplcheck
+=========
+
+``chplcheck`` is a linter for the Chapel programming language implemented in
+Python using the :ref:`Python bindings for the compiler frontend <readme-chapel-py>`.
+It is intended to catch stylistic mistakes and bad practices in Chapel programs.
+It is also intended to be customizable and extensible, using a system of named
+'rules' that lead to warnings.
+
+``chplcheck`` supports the Language Server Protocol, allowing it to be used as
+part of your favorite editor. The following image demonstrates its' use in
+Neovim:
+
+.. image:: neovim.png
+  :alt: Screenshot of code using ``chplcheck``
+
+Getting Started
+---------------
+
+The easiest way to make ``chplcheck`` available on your command line is by using the
+``chplcheck`` Makefile target. This will build the Dyno compiler frontend and the
+Python bindings for Dyno if needed, and place ``chplcheck`` into ``$CHPL_HOME/bin``.
+In from-source builds, this directory should already be in your ``PATH``;
+this means you should be invoke ``chplcheck`` from the command-line without any
+additional work.
+
+.. code-block:: bash
+
+   cd $CHPL_HOME
+   make chplcheck
+   chplcheck --help
+
+Saving the following file into ``myfile.chpl``:
+
+.. code-block:: chapel
+   :linenos:
+
+   record MyRecord {}
+
+   for i in 1..10 do {
+     writeln("Hello, world!");
+   }
+
+The linter is run as follows:
+
+.. code-block:: bash
+
+   > chplcheck myfile.chpl
+   path/to/myfile/myfile.chpl:1: node violates rule CamelCaseRecords
+   path/to/myfile/myfile.chpl:3: node violates rule DoKeywordAndBlock
+   path/to/myfile/myfile.chpl:3: node violates rule UnusedLoopIndex
+
+
+Enabling / Disabling Rules
+--------------------------
+
+Each rule, such as ``CamelCaseRecords``, can be individually enabled or
+disabled from the command line using ``--enable-rule`` and ``--disable-rule``.
+To silence the warning about unused loop indices such as ``i`` in the above
+code, we can invoke ``chplcheck`` as follows:
+
+.. code-block:: bash
+
+   > chplcheck myfile.chpl --disable-rule UnusedLoopIndex
+   path/to/myfile/myfile.chpl:1: node violates rule CamelCaseRecords
+   path/to/myfile/myfile.chpl:3: node violates rule DoKeywordAndBlock
+
+
+Some rules are disabled by default. One such rule is ``UseExplicitModules``, which
+warns against letting Chapel automatically create the top-level module in a file.
+
+.. code-block:: bash
+
+   > chplcheck myfile.chpl --enable-rule UseExplicitModules
+   path/to/myfile/myfile.chpl:1: node violates rule CamelCaseRecords
+   path/to/myfile/myfile.chpl:1: node violates rule UseExplicitModules
+   path/to/myfile/myfile.chpl:3: node violates rule DoKeywordAndBlock
+   path/to/myfile/myfile.chpl:3: node violates rule UnusedLoopIndex
+
+
+All of the rules present in ``chplcheck`` right now are defined in
+``tools/chplcheck/src/rules.py``. You may reference this file to get a complete
```

Comment:
This is meh and I've opened a (self-assigned) issue to fix it. https://github.com/chapel-lang/chapel/issues/24622